wasm/section.go

```diff
@@ -105,19 +105,45 @@ func (e MissingSectionError) Error() string {
 	return fmt.Sprintf("wasm: missing section %s", SectionID(e).String())
 }
 
+type sectionsReader struct {
+	lastSecOrder uint8 // previous non-custom sectionid
+	m            *Module
+}
+
+func newSectionsReader(m *Module) sectionsReader {
+	return sectionsReader{m: m}
+}
+
+func (s *sectionsReader) readSections(r *readpos.ReadPos) error {
+	for {
+		done, err := s.readSection(r)
+		if err != nil {
+			return err
+		} else if done {
+			return nil
+		}
+	}
+}
+
 // reads a valid section from r. The first return value is true if and only if
 // the module has been completely read.
-func (m *Module) readSection(r *readpos.ReadPos) (bool, error) {
-	var err error
-	var id uint32
+func (sr *sectionsReader) readSection(r *readpos.ReadPos) (bool, error) {
+	m := sr.m
 
 	logger.Println("Reading section ID")
-	id, err = leb128.ReadVarUint32(r)
+	id, err := r.ReadByte()
 	if err == io.EOF {
 		return true, nil
 	} else if err != nil {
 		return false, err
 	}
+	if id != uint8(SectionIDCustom) {
+		if id <= sr.lastSecOrder {
+			return false, fmt.Errorf("wasm: sections must occur at most once and in the prescribed order")
+		}
+		sr.lastSecOrder = id
+	}
+
 	s := RawSection{ID: SectionID(id)}
```

Comment:
reading out the section id as uint32 and truncating to uint8 will cause problem.